### PR TITLE
rt refine: sorries in CNodeInv and Ipc

### PR DESCRIPTION
**Pull request by pierzchalski**
For whatever reason, these proofs are sensitive to the internals of `cancelIPC`, which were updated in my branch `eap/replyRemoveTCB_invs'` - hence that being the target.

These are the proofs that @MitchellBuckley and @michaelmcinerney handed off to me last week.

### CODE REVIEW

File: proof/refine/ARM/CNodeInv_R.thy [hunk 1]
```diff
@@ -7026,20 +7092,26 @@ next
            apply (clarsimp simp: capRemovable_def cte_wp_at_ctes_of)
           apply (clarsimp dest!: isCapDs simp: cte_wp_at_ctes_of)
           apply (case_tac "cteCap rv'",
-                 auto simp add: isCap_simps is_cap_simps final_matters'_def)[1]
-         apply (wp isFinalCapability_inv static_imp_wp
-                 | simp add: is_final_cap_def conj_comms cte_wp_at_eq_simp)+
-       apply (rule isFinal[where x="cte_map slot"])
-      apply (wp get_cap_wp| simp add: conj_comms)+
-      apply (wp getCTE_wp')
+                 auto simp add: isCap_simps is_cap_simps final_matters'_def pred_tcb_at'_def
```
Comment (michaelmcinerney) [inline proof/refine/ARM/CNodeInv_R.thy:7095]: Is the auto necessary here?

**Review by michaelmcinerney — APPROVED** [review]

File: proof/refine/ARM/CNodeInv_R.thy (hunk 1 above)
Comment (pierzchalski) [reply proof/refine/ARM/CNodeInv_R.thy:7095]: Weirdly, yes (you can see its predecessor at line 7029 above). IIRC there's a schematic involved that `force` and friends don't want to instantiate, but `auto` is happy to.

Comment (michaelmcinerney) [reply proof/refine/ARM/CNodeInv_R.thy:7095]: Blerg

**Review by corlewis — APPROVED** [review]

File: proof/refine/ARM/Ipc_R.thy [hunk 2]
```diff
@@ -2811,36 +2790,30 @@ lemma cteDeleteOne_ResumeCurrentThread_imp_notct:
...
+        apply (wpc)
+         apply (wp | simp)+
+      apply (wpc, wp+)
+    apply (rule_tac Q="\<lambda>_. ?PRE t'" in hoare_post_imp, clarsimp)
+    apply (wpsimp wp: stateAssert_wp)+
   done
-qed
+
+lemma blockedCancelIPC_ResumeCurrentThread_imp_notct[wp]:
+  "blockedCancelIPC a b c \<lbrace>\<lambda>s. ksSchedulerAction s = ResumeCurrentThread \<longrightarrow> ksCurThread s \<noteq> t'\<rbrace>"
```
Comment (MitchellBuckley) [inline proof/refine/ARM/Ipc_R.thy:2809]: nitpick: rename a b c to something more meaningful

File: proof/refine/ARM/KHeap_R.thy [hunk 3]
```diff
@@ -2534,6 +2573,25 @@ lemma setReply_state_refs_of'[wp]:
...
+   apply (clarsimp simp: ext opt_map_def list_refs_of_reply'_def map_set_def projectKO_opt_reply
+                  split: option.splits)+
+  done
+
+lemma updateReply_wp_all:
+  "\<lbrace>\<lambda>s. \<forall>ko. ko_at' ko rptr s \<longrightarrow> P (set_obj' rptr (upd ko) s)\<rbrace>
+   updateReply rptr upd
+   \<lbrace>\<lambda>_. P\<rbrace>"
+  unfolding updateReply_def
+  apply (wpsimp wp: set_reply'.set_wp)
```
Comment (MitchellBuckley) [inline proof/refine/ARM/KHeap_R.thy:2592]: nitpick: use by in these very short proofs. This is a true nitpick, feel free to ignore.

File: proof/refine/ARM/IpcCancel_R.thy [hunk 4]
```diff
@@ -960,8 +958,8 @@ lemma cancelAllIPC_tcb_at_runnable':
 lemma cancelAllSignals_tcb_at_runnable':
   "\<lbrace>st_tcb_at' runnable' t\<rbrace> cancelAllSignals ntfnptr \<lbrace>\<lambda>_. st_tcb_at' runnable' t\<rbrace>"
   unfolding cancelAllSignals_def
-  sorry (*
-  by (wpsimp wp: mapM_x_wp' sts_st_tcb' hoare_drop_imp) *)
+  oops
```
Comment (MitchellBuckley) [inline proof/refine/ARM/IpcCancel_R.thy:961]: please add a comment explaining why this is an oops and not a sorry.

**Review by MitchellBuckley — APPROVED** [review]

File: proof/refine/ARM/CNodeInv_R.thy [hunk 5]
```diff
@@ -7035,20 +7102,26 @@ next
...
-    apply (clarsimp simp: cte_wp_at_ctes_of)
-    apply (frule ctes_of_valid', clarsimp)
-    apply ((clarsimp | rule conjI)+)[1]
-    done *)
+     apply (frule(1) cte_wp_at_valid_objs_valid_cap[where P="(=) cap" for cap,
+                                                    OF _ invs_valid_objs])
+     supply split_paired_Ex[simp del]
+     apply (clarsimp simp: invs_def valid_state_def valid_pspace_def cte_wp_at_def)
+     apply (subst split_paired_Ex[symmetric])
+     apply auto[1]
```
[the remark this answers is not in the document]
Comment (pierzchalski) [reply proof/refine/ARM/CNodeInv_R.thy:7119]: ah, `subgoal by auto` would be better, yes!

[the remark this answers is not in the document]
Comment (pierzchalski) [reply proof/refine/ARM/CNodeInv_R.thy:7119]: Oh, looks like `subgoal` changes the goal in a way that prevents `auto` from making progress with unification. I've replaced this with `apply (solves "auto")[1]` to get a similar effect.

File: proof/refine/ARM/CNodeInv_R.thy [hunk 6]
```diff
@@ -6198,10 +6264,10 @@ proof (induct arbitrary: P p rule: finalise_spec_induct2)
         apply clarsimp
        apply (clarsimp simp: cte_wp_at_ctes_of capRemovable_def)
        apply (subgoal_tac "final_matters' (cteCap rv) \<and> \<not> isUntypedCap (cteCap rv)")
-        apply clarsimp
-        apply (rule conjI)
-         apply clarsimp
-        apply clarsimp
+        apply (intro conjI impI
```
[the remark this answers is not in the document]
Comment (pierzchalski) [reply proof/refine/ARM/CNodeInv_R.thy:6267]: Just had another go, if so it's not immediately obvious. It's frustratingly common for `fastforce` to be very fast on the subgoals that come out of `intro`, but very slow on the original subgoal.

**Comment by pierzchalski** [comment]
[Testboard](https://bamboo.ts.data61.csiro.au/browse/L4V-TESTBOARD-637).

**Comment by pierzchalski** [comment]
[Post-rebase testboard](https://bamboo.ts.data61.csiro.au/browse/L4V-TESTBOARD-646)

**Review by mktnk3 — APPROVED** [review]
This looks good! Getting close to 2 digit count... :)

**Pull request merged**